### DISCUSSION TITLE
unable to write to mergerfs pool

### DISCUSSION THREAD

**Discussion by spediegunz** [trapexit/mergerfs#1222] [Q&A]
hello with the help of chatgpt ive been trying to mimic a setup a had provided by a now no longer supported software called plexguide. essentially it used mergerfs to pool together google drive and a local folder. because of the rising prices of cloud storage i purchased a nas and I'm trying to configure mergerfs to do the same thing since large 4k files take quite some time to be copied over to the nas and I want plex to have access to them as soon as possible.  I created this pool in my fstab

mergerfs#/mnt/local=RW:/mnt/media=NC /mnt/pcnasmerge fuse.mergerfs direct_io,defaults,sync_read,auto_cache,dropcacheonclose=true,use_ino,allow_other,func.getattr=newest,category.create=ff,minfreespace=0 0 0

the goal is to always make new files be created on the /mnt/local branch and a separate script would copy them over to the nas and delete the originals when the copy is complete. the first day i set this up and sonarr and radarr had no issues downloading files to the merger pool /mnt/pcnasmerge but im not sure what has since happened all my downloads cannot be imported to this path

2023-8-6 07:29:48.1|Warn|ImportApprovedMovie|Couldn't import movie /mnt/unionfs/sabnzbd/The Big Four 2022 1080p WEB h264-KOGi/The.Big.Four.2022.1080p.WEB.h264-KOGi.mkv

[v3.0.2.4552] System.IO.DirectoryNotFoundException: Could not find a part of the path '/mnt/pcnasmerge/movies/The Big 4 (2022)/The Big 4 (2022) WEBDL-1080p.mkv'.
   at System.IO.FileSystem.LinkOrCopyFile(String sourceFullPath, String destFullPath) 

before i started trying to sort out what was going on with those apps i simply tried to just use the command line to create a directory in the merger pool directory and i keep on getting this error
ronald@homeserver:/mnt/pcnasmerge$ sudo mkdir testingpurposes
mkdir: cannot create directory ‘testingpurposes’: Read-only file system

chatgpt is claiming that i need both branches to be rw in order for the merger path to also be rw but don't think this is correct because the merger pool made by plexguide for the google drive unionfs uses these same rules and work fine can anyone provide me with guidance on how to correctly get this working as intended.

**Comment by trapexit** [comment]
https://github.com/trapexit/mergerfs#filtering

mergerfs only returns what the OS tells it to or in the case it filters out everything ENOSPC or ENOENT or EROFS as shown in the docs. 

You told it to search in order. /mnt/local must be being filtered out and because you said /mnt/cloud=NC it returns EROFS because that was the last error.

You have to look at your setup and figure out why /mnt/local is skipped.

I can't help you further without more info: https://github.com/trapexit/mergerfs#support

**Comment by spediegunz** [comment]
you might be correct because i notice if i change /mnt/media to rw leaving first found on it wouldn't give a read only error how ever it would still only write to the /mnt/media branch despite being listed second, same thing when I use lfs which is the /mnt/local path its still writes to /mnt/media so for some reason its filtering out or skipping /mnt/local. 

the only thing that has changed since it was working correctly was i implemented script to copy the contents of /mnt/local to /mnt/media and delete the originals and a service and timer so it executes every 4 hours is it possible that the service might be doing something to the /mnt/local path causing it to be skipped.  Also I should mention I am able to create files directly to the /mnt/local path.

```
**here is the script**
#!/bin/bash

# Function to copy files from source to destination
copy_files() {
    SOURCE="$1"
    DESTINATION="$2"

    echo "$(date) - Copying files from: $SOURCE to $DESTINATION"

    # If the source is a directory, copy the entire directory structure to the destination
    if [ -d "$SOURCE" ]; then
        sudo rsync -ah --no-o --no-g --no-perms "$SOURCE"/ "$DESTINATION"/
    fi
}

# Main script starts here

# Path to the source folder (where new movie files are added)
SOURCE_FOLDER="/mnt/local/movies"
# Destination folder
DESTINATION_FOLDER="/mnt/media/movies"

# Run the copy_files function for the entire source folder
copy_files "$SOURCE_FOLDER" "$DESTINATION_FOLDER"

# Delete copied files and empty directories in the source folder
find "$SOURCE_FOLDER" -mindepth 1 -delete
```

```
**here is the service**  
[Unit]
Description=Copy and Organize 4kMovies

[Service]
Type=simple
User=ronald
WorkingDirectory=/mnt/scripts
ExecStart=/bin/bash /mnt/scripts/copy_and_organize_4kmovies.sh

[Install]
WantedBy=default.target


```
its a long shot because ive tried disabling all scripts and services but I still can't write to /mnt/pcnasmerge.  Can you tell me the best course of action to figure out what could be causing this issue ?

**Reply by trapexit** [reply]
As I mentioned above it would be due to filtering. So look at your settings and see what attributes could be causing it to be filtered. Did you change any values without remounting / rebooting? What does the runtime API say the values are?

**Comment by spediegunz** [comment]
for testing purposes i unmounted the /mnt/pcnasmerge and commented out the fstab entry then i manually mounted it again using this command

`sudo mount -t fuse.mergerfs -o defaults,allow_other,use_ino,category.create=ff,minfreespace=0 /mnt/local:/mnt/media /mnt/pcnasmerge`

everything worked perfectly i was able to create files and folders in the /mnt/pcnasmerge directory and the files went straight to the /mnt/local folder as expected behavior, afterwards i restarted my docker containers for sonarr and radarr and the movies and tv shows imported without a hitch to the /pcnasmerge directory.  

afterward I unmounted again and simplified my fstab entry to shadow the manual mount i just tested and confirmed to work correctly

`mergerfs#/mnt/local:/mnt/media /mnt/pcnasmerge fuse.mergerfs defaults,allow_other,use_ino,category.create=ff,minfreespace=0 0 0
`
`sudo mount -a
`
unfortunately the same result all files created in /mnt/pcnasmerge went straight to the /mnt/media path despite the ff create policy.

something going on with the fstab causing the filtering issues

**Reply by trapexit** [reply]
Why do you have `mergerfs#/mnt/local:/mnt/media`

The value should just be `/mnt/local:/mnt/media`

**Reply by spediegunz** [reply]
wow, that seems like that was the problem after I edited it to the way you wrote the merger pool wrote directly to the first found local path, thank you for your help Ill finish trying to configure everything along the scripts and sonarr and radarr to make sure everything is working as desired.  

as for why was the fstab entry written the wrong way, well i guess chatgpt still has some learning to do lol

**Reply by trapexit** [reply]
You must always double check anything from a LLM. I work in ML for my day job and today I'm actually testing Llama 2. You can't trust anything that comes from them. Hell... ChatGPT gets something as simple as the author of mergerfs wrong at least 50% of the time and I don't think I've ever seen it give the license correctly.

**Comment by spediegunz** [comment]
because my local network speeds are quite slow, ive decided to use mergerfs to help make media available to plex faster for playback rather than wait the long times for files to be moved over to the nas after being downloaded 

for context I want to lay out my order of processes 

1. sonarr and radarr download to the mergerfs pool directory which is a merger of my local download path and my nas here is the fstab entry for that mount 

`/mnt/local=RW:/mnt/media=NC: /mnt/pcnasmerge fuse.mergerfs direct_io,defaults,sync_read,auto_cache,dropcacheonclose=true,use_ino,allow_other,func.getattr=newest,category.create=ff,minfreespace=0 0 0`

2. plex can see and the files and play them directly off of the merger path without any delay because i can read them off of the local branch 

3. every few hours a script runs that copies the directories from the local branch to the nas branch of the merger pool and deletes the originals from the local branch 


after a few weeks of testing, my setup is working almost exactly as intended, however I'm still having a problem with sonarr.  when i add a new show sonarr creates a new directory in the merger pool path and everything works perfectly. The issue only happens when it needs to add an episode to an already existing show directory in the merger pool that has since been copied over to the nas branch from the regularly scheduled script.  In the eyes of sonarr nothing should of changed because the shows directories are still visible in the merger pool path even though they were copied to the second branch but for some reason it can tell something is off and this is the error I get.  I don't believe I'm having a permissions issue because I have no problems downloading new or existing shows when setting the download path in sonarr directly to /mnt/media (my nas branch) or to /mnt/local(my local branch) also ive opened all permissions to any user or group.

I know this might be more of a sonarr issue but i thought i would ask here because maybe I'm missing something with the merger pool mount that needs to be added so sonarr wont run into problems.

> 23-8-17 16:37:02.6|Warn|ImportApprovedEpisodes|Couldn't import episode /mnt/downloads/sabnzbd/Futurama S01 MULTiSUBS 1080p WEB H264-DRAUGR/Futurama.S01E03.I.Roommate.1080p.WEB.H264-DRAUGR.mkv
> 
> > [v3.0.5.1143] System.UnauthorizedAccessException: Access to the path "/mnt/pcnasmerge/tv/Futurama/Season 1" is denied.
> >   at System.IO.Directory.CreateDirectoriesInternal (System.String path) [0x0005e] in <254335e8c4aa42e3923a8ba0d5ce8650>:0 
> >   at System.IO.Directory.CreateDirectory (System.String path) [0x0008f] in <254335e8c4aa42e3923a8ba0d5ce8650>:0 
> >   at NzbDrone.Common.Disk.DiskProviderBase.CreateFolder (System.String path) [0x00047] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Common\Disk\DiskProviderBase.cs:190 
> >   at NzbDrone.Core.MediaFiles.EpisodeFileMovingService.CreateFolder (System.String directoryName) [0x00079] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Core\MediaFiles\EpisodeFileMovingService.cs:203 
> >   at NzbDrone.Core.MediaFiles.EpisodeFileMovingService.EnsureEpisodeFolder (NzbDrone.Core.MediaFiles.EpisodeFile episodeFile, NzbDrone.Core.Tv.Series series, System.Int32 seasonNumber, System.String filePath) [0x00098] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Core\MediaFiles\EpisodeFileMovingService.cs:173 
> >   at NzbDrone.Core.MediaFiles.EpisodeFileMovingService.EnsureEpisodeFolder (NzbDrone.Core.MediaFiles.EpisodeFile episodeFile, NzbDrone.Core.Parser.Model.LocalEpisode localEpisode, System.String filePath) [0x00000] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Core\MediaFiles\EpisodeFileMovingService.cs:146 
> >   at NzbDrone.Core.MediaFiles.EpisodeFileMovingService.MoveEpisodeFile (NzbDrone.Core.MediaFiles.EpisodeFile episodeFile, NzbDrone.Core.Parser.Model.LocalEpisode localEpisode) [0x00026] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Core\MediaFiles\EpisodeFileMovingService.cs:75 
> >   at NzbDrone.Core.MediaFiles.UpgradeMediaFileService.UpgradeEpisodeFile (NzbDrone.Core.MediaFiles.EpisodeFile episodeFile, NzbDrone.Core.Parser.Model.LocalEpisode localEpisode, System.Boolean copyOnly) [0x001ab] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Core\MediaFiles\UpgradeMediaFileService.cs:77 
> >   at NzbDrone.Core.MediaFiles.EpisodeImport.ImportApprovedEpisodes.Import (System.Collections.Generic.List`1[T] decisions, System.Boolean newDownload, NzbDrone.Core.Download.DownloadClientItem downloadClientItem, NzbDrone.Core.MediaFiles.EpisodeImport.ImportMode importMode) [0x0029b] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Core\MediaFiles\EpisodeImport\ImportApprovedEpisodes.cs:109

**Reply by trapexit** [reply]
1) Your settings look old (using deprecated settings) and contradictory. You can't have direct_io and not (autocache) at the same time.
2) Just because on the surface you seem to think they are the same doesn't mean they are. There is metadata that changes. The inode for instance, timestamp maybe, etc.
3) If you change things out of band like that and you have caching enabled you will run into weird issues at times. As I mention in the docs. You can't do caching and change things out of band. That's asking for trouble.
